include/lattice_field.h

```diff
@@ -416,25 +468,72 @@ namespace quda {
     mutable bool backed_up;
 
   public:
+    /**
+       Static variable that is determined which ghost buffer we are using
+     */
+    inline static int bufferIndex = 0;
 
     /**
-       Constructor for creating a LatticeField from a LatticeFieldParam
-       @param param Contains the metadata for creating the LatticeField
+       @brief Default constructor
     */
-    LatticeField(const LatticeFieldParam &param);
+    LatticeField();
 
     /**
-       Constructor for creating a LatticeField from another LatticeField
-       @param field Instance of LatticeField from which we are
-       inheriting metadata
+       @brief Copy constructor for creating a LatticeField from another LatticeField
+       @param field Instance of LatticeField from which we are cloning
     */
     LatticeField(const LatticeField &field);
 
     /**
-       Destructor for LatticeField
+       @brief Move constructor for creating a LatticeField from another LatticeField
+       @param field Instance of LatticeField from which we are moving
+    */
+    LatticeField(LatticeField &&field);
+
+    /**
+       @brief Constructor for creating a LatticeField from a LatticeFieldParam
+       @param param Contains the metadata for creating the field
+    */
+    LatticeField(const LatticeFieldParam &param);
+
+    /**
+       @brief Destructor for LatticeField
     */
     virtual ~LatticeField();
-    
+
+    /**
+       @brief Copy assignment operator
+       @param[in] field Instance from which we are copying
+       @return Reference to this field
+     */
+    LatticeField &operator=(const LatticeField &);
+
+    /**
+       @brief Move assignment operator
+       @param[in] field Instance from which we are moving
+       @return Reference to this field
+     */
+    LatticeField &operator=(LatticeField &&);
+
+    /**
+       @brief Create the field as specified by the param
+       @param[in] Parameter struct
+    */
+    void create(const LatticeFieldParam &param);
```

Comment:
Do we still need/want to have `create`/`clear` ? Is there any benefit over construct/destruct properly?

Reply:
I added the `create` function to minimize the boilerplate, and make things less error prone by reusing the same code.  Though I guess I should make `create` private so it's clear that this cannot be called outside of the class.

`clear` is aping `std` containers, and is required where we have file-scope variables to prevent memory leaks.  This is especially evident in some of the tests for example.

Reply:
So clear is needed when variables don't go out of scope before `endQuda` ?
When we just had pointers calling delete made sense but file-scope variables will go out of scope and having to call clear on them is not what one would expect.
Making `create` private is certainly good.

Reply:
Yes, `clear` is needed when variables don't go out of scope before `endQuda`.  The situation is the same if one created  a `std::vector` but had it use QUDA's allocators.

Reply:
So this is some unexpected behavior. When I have a variable going out of scope before endQuda is called I don't need to call clear but if it doesn't I have to. I think that is a confusing interface. I'd prefer if we could come up with a better solution for this. Would it be sufficient to call `endQuda` in the tests after all variables go out of scope to avoid that?
Maybe we could have a smart pointer to a C++ Quda handle that calls `endQuda` in a custom deleter? For C code that just use the C interface everything can stay as is as they cannot use QUDA C++ objects anyway.

Reply:
The variables in the tests don't go out of scope ever, since they have file scope.  So given that, this isn't that surprising I think.

We could rewrite the tests to avoid this, but that is an overhaul that goes well beyond the scope (pun not intended) of this PR.

Reply:
a285361b8900d516da4af9bb79237bcd6adf1c58 (create now private)